pydra/engine/tests/test_shelltask.py

```diff
@@ -2450,6 +2479,298 @@ def test_shell_cmd_outputspec_wf_1(plugin):
+        name="Output",
+        fields=[
+            (
+                "newfile1",
+                File,
+                {
+                    "output_file_template": "{file1}",
+                    "help_string": "newfile 1",
+                    "requires": ["file1"],
+                },
+            ),
+            (
+                "newfile2",
+                File,
+                {
+                    "output_file_template": "{file2}",
+                    "help_string": "newfile 1",
+                    "requires": ["file1", "file2"],
+                },
+            ),
+        ],
+        bases=(ShellOutSpec,),
+    )
+    shelly = ShellCommandTask(
+        name="shelly",
+        executable=cmd,
+        input_spec=my_input_spec,
+        output_spec=my_output_spec,
+    )
+    shelly.inputs.file1 = "new_file_1.txt"
+
+    res = shelly()
+    assert res.output.stdout == ""
+    assert res.output.newfile1.exists()
+    assert res.output.newfile2 is attr.NOTHING
+
+
+def test_shell_cmd_inputspec_outputspec_3():
+    """
+    customised input_spec and output_spec, output_spec uses input_spec fields in the requires filed
+    adding one additional input that is not in the template, but in the requires field,
+    """
+    cmd = ["touch", "newfile_tmp.txt"]
+    my_input_spec = SpecInfo(
+        name="Input",
+        fields=[
+            (
+                "file1",
+                str,
+                {"help_string": "1st creadted file", "argstr": "", "position": 1},
+            ),
+            (
+                "file2",
+                str,
+                {"help_string": "2nd creadted file", "argstr": "", "position": 2},
+            ),
+            ("additional_inp", str, {"help_string": "additional inp"}),
+        ],
+        bases=(ShellSpec,),
+    )
+
+    my_output_spec = SpecInfo(
+        name="Output",
+        fields=[
+            (
+                "newfile1",
+                File,
+                {
+                    "output_file_template": "{file1}",
+                    "help_string": "newfile 1",
+                    "requires": ["file1"],
+                },
+            ),
+            (
+                "newfile2",
+                File,
+                {
+                    "output_file_template": "{file2}",
+                    "help_string": "newfile 1",
+                    "requires": ["file1", "file2", "additional_inp"],
+                },
+            ),
+        ],
+        bases=(ShellOutSpec,),
+    )
+    shelly = ShellCommandTask(
+        name="shelly",
+        executable=cmd,
+        input_spec=my_input_spec,
+        output_spec=my_output_spec,
+    )
+    shelly.inputs.file1 = "new_file_1.txt"
+    shelly.inputs.file2 = "new_file_2.txt"
+
+    res = shelly()
+    assert res.output.stdout == ""
+    assert res.output.newfile1.exists()
+    # additional input not provided so no newfile2 set (even if the file was created)
+    assert res.output.newfile2 is attr.NOTHING
```

Comment:
should there be an additional check that when `addtional_inp` is provided that it is generated? 

also should there be a check/proper exception raised on the input/outputspec if someone uses a `output_file_template` which has an arg that is not defined in the inputs?  this should help users debug the specs better. perhaps a check_specs function that can be used in standalone mode by a user, but also during runtime.

Reply:
will add a tests that shows that it works if `additional_inp` provided.

For the second question - i thought that we want to treat is as an "optional", so simply do not return the output when specific set of the input is not set. 
Or do you mean, that fields i the `requires` should be treated like this, but field that are not in the `requires` but only in the template should lead to the error if they are not set?